bin/bnf_to_cnf/bnf_to_cnf/driver.py

```diff
@@ -83,38 +86,42 @@ def merge(self, driver: "Driver"):
         self.tree.merge(driver.tree)
 
 
-def load_script(filename: str, cache: Dict[str, Driver] = dict()):
+def load_script(filepath: Path, cache: Dict[str, Driver] = dict()):
     """Recursively load a script, parsing it and adding dependencies.
 
     Args:
-        filename: The name of the file to open.
+        filepath: The path of the file to open.
         cache: A cache to avoid duplicate work.
 
     Returns:
         The fully parsed grammar.
 
     """
+    filepath = filepath.resolve()
+    filename = str(filepath)
+
     assert filename not in cache
```

Comment:
Not sure about this line. I think it isn't needed and doesn't make for a good API.

Reply:
Hm, `filename` is a misleading variable name, since it's going to contain the full file path (and not just the file name) as a string.

Maybe:
```python
filepath_str = str(filepath.resolve())

assert filepath_str not in cache
```

Reply:
That's a good idea. I was commenting on the assert statement though, github might not display that clearly.

Reply:
Ah, you're right, there shouldn't be assertions in production code.

If I'm getting the purpose of the assert right, the `cache` passed into the function may contain Python files (recursively) imported by the given file, but must not contain the file itself?

I would indeed replace the `assert` with an `if ... raise`.

Reply:
By the way, the default value in `cache: Dict[str, Driver] = dict()` is dangerous. If a default value is needed for this function, I would use
```python
def load_script(filepath: Path, cache: Optional[Dict[str, Driver]] = None):
    if cache is None:
        cache = {}
```

Reply:
> the default value in `cache: Dict[str, Driver] = dict()` is "dangerous"

This is on purpose because the cache is/must be shared between function calls.

Reply:
> If I'm getting the purpose of the assert right, the `cache` passed into the function may contain Python files (recursively) imported by the given file, but must not contain the file itself?

Each import adds the productions for the imported grammar file. Because productions shouldn't be duplicated to reduce runtime we have to check whether a file was imported already.